lib/nerve.rb

```diff
@@ -61,107 +69,118 @@ def load_config!
+                # Keep the old watcher running until the replacement is launched
+                # This keeps the service registered while we change it over
+                # This also keeps connection pools active across diffs
+                temp_name = "#{name}_#{@watcher_versions[name]}"
+                @watchers[temp_name] = @watchers.delete(name)
+                @watcher_versions[temp_name] = @watcher_versions.delete(name)
+                log.info "nerve: launching new watcher for #{name}"
+                launch_watcher(name, @watchers_desired[name], :wait => true)
+                log.info "nerve: reaping old watcher #{temp_name}"
+                reap_watcher(temp_name)
+              end
             end
-          end
 
-          # If this was a configuration check, bail out now
-          if @config_manager.options[:check_config]
-            log.info 'nerve: configuration check succeeded, exiting immediately'
-            break
-          end
+            # If this was a configuration check, bail out now
+            if @config_manager.options[:check_config]
+              log.info 'nerve: configuration check succeeded, exiting immediately'
+              break
+            end
 
-          # Check that watchers are still alive, auto-remediate if they
-          # are not. Sometimes zookeeper flakes out or connections are lost to
-          # remote datacenter zookeeper clusters, failing is not an option
-          relaunch = []
-          @watchers.each do |name, watcher|
-            unless watcher.alive?
-              relaunch << name
+            # Check that watchers are still alive, auto-remediate if they
+            # are not. Sometimes zookeeper flakes out or connections are lost to
+            # remote datacenter zookeeper clusters, failing is not an option
+            relaunch = []
+            @watchers.each do |name, watcher|
+              unless watcher.alive?
+                relaunch << name
+              end
             end
-          end
 
-          relaunch.each do |name|
-            begin
-              log.warn "nerve: watcher #{name} not alive; reaping and relaunching"
-              reap_watcher(name)
-            rescue => e
-              log.warn "nerve: could not reap #{name}, got #{e.inspect}"
+            relaunch.each do |name|
+              begin
+                log.warn "nerve: watcher #{name} not alive; reaping and relaunching"
+                reap_watcher(name)
+                statsd.increment('nerve.watcher.reaping', tags: ['reap_result:success', "watcher_name:#{name}"])
+              rescue => e
+                log.warn "nerve: could not reap #{name}, got #{e.inspect}"
+                statsd.increment('nerve.watcher.reaping', tags: ['reap_result:fail', "watcher_name:#{name}"])
+              end
+              launch_watcher(name, @watchers_desired[name])
+              statsd.increment('nerve.watcher.relaunch', tags: ["watcher_name:#{name}"])
```

Comment:
maybe to have this before the launch_watcher call? so that it can emit a trace even if launch_watcher fails